File: src/main/java/net/nodus/core/site/SiteService.java
package net.nodus.core.site;

import java.util.List;
import net.nodus.global.common.exception.GlobalException;
import net.nodus.core.site.dto.CreateSiteRequest;
import net.nodus.core.site.dto.IssuedSiteKey;
import net.nodus.core.site.dto.SiteCreateResult;
import net.nodus.core.site.dto.UpdateSiteRequest;
import net.nodus.database.site.SiteKeyRepository;
import net.nodus.database.site.SiteKeyStatus;
import net.nodus.database.site.SiteRepository;
import net.nodus.database.site.entity.Site;
import net.nodus.database.site.entity.SiteKey;
import org.springframework.stereotype.Service;

@Service
public class SiteService {

    private final SiteRepository siteRepository;
    private final SiteKeyRepository siteKeyRepository;
    private final SiteKeyService siteKeyService;

    public SiteService(SiteRepository siteRepository, SiteKeyRepository siteKeyRepository,
        SiteKeyService siteKeyService) {
        this.siteRepository = siteRepository;
        this.siteKeyRepository = siteKeyRepository;
        this.siteKeyService = siteKeyService;
    }

    public SiteCreateResult create(String userId, CreateSiteRequest request) {
        Site site = siteRepository.save(
            new Site(userId, request.getName(), request.getDomain(), request.getUrl()));
        return new SiteCreateResult(site, siteKeyService.issue(site));
    }

    public List<Site> list(String userId) {
        return siteRepository.findByUserAccountIdAndDeletedAtIsNullOrderByCreatedAtDesc(userId);
    }

    public Site get(String userId, String siteId) {
        return siteRepository.findByIdAndUserAccountIdAndDeletedAtIsNull(siteId, userId)
            .orElseThrow(() -> new GlobalException.DataNotFound("Site not found"));
    }

    public Site update(String userId, String siteId, UpdateSiteRequest request) {
        Site site = get(userId, siteId);

        if (request.getName() != null) {
            site.setName(request.getName());
        }
        site.setDomain(request.getDomain());
        site.setUrl(request.getUrl());
        site.touch();

        return siteRepository.save(site);
    }

    public void delete(String userId, String siteId) {
        Site site = get(userId, siteId);

        site.softDelete();

        siteRepository.save(site);
        siteKeyService.revokeActive(siteId);
    }

    public SiteKey keyInfo(String userId, String siteId) {
        get(userId, siteId);

        return siteKeyRepository.findBySiteIdAndStatusAndDeletedAtIsNull(siteId,
                SiteKeyStatus.ACTIVE)
            .orElseThrow(() -> new GlobalException.DataNotFound("Active site key not found"));
    }

    public IssuedSiteKey rotateKey(String userId, String siteId) {
        Site site = get(userId, siteId);
        return siteKeyService.issue(site);
    }

    public void revokeKey(String userId, String siteId) {
        get(userId, siteId);
        siteKeyService.revokeActive(siteId);
    }
}

File: src/main/kotlin/net/nodus/site/SiteService.kt
package net.nodus.site

import net.nodus.config.exception.GlobalException
import net.nodus.project.ProjectRepository
import net.nodus.site.dto.CreateSiteRequest
import net.nodus.site.dto.IssuedSiteKey
import net.nodus.site.dto.SiteCreateResult
import net.nodus.site.dto.UpdateSiteRequest
import net.nodus.site.entity.Site
import net.nodus.site.entity.SiteKey
import org.springframework.stereotype.Service
import java.time.Instant

@Service
class SiteService(
    private val siteRepository: SiteRepository,
    private val siteKeyRepository: SiteKeyRepository,
    private val siteKeyService: SiteKeyService,
    private val projectRepository: ProjectRepository,
) {

    fun create(userId: String, request: CreateSiteRequest): SiteCreateResult {
        val project = projectRepository.findByIdAndUserAccountIdAndDeletedAtIsNull(
            id = request.projectId,
            userAccountId = userId,
        ) ?: throw GlobalException.DataNotFound("Project not found.")

        val site = siteRepository.save(
            Site(
                userAccountId = userId,
                workspaceId = project.workspaceId,
                projectId = requireNotNull(project.id),
                name = request.name,
                domain = request.domain,
                url = request.url,
            )
        )

        return SiteCreateResult(site = site, issuedKey = siteKeyService.issue(site))
    }

    fun list(userId: String, projectId: String): List<Site> =
        siteRepository.findByUserAccountIdAndProjectIdAndDeletedAtIsNullOrderByCreatedAtDesc(
            userAccountId = userId,
            projectId = projectId
        )

    fun get(userId: String, siteId: String): Site =
        siteRepository.findByIdAndUserAccountIdAndDeletedAtIsNull(siteId, userId)
            ?: throw GlobalException.DataNotFound("Site not found")

    fun update(userId: String, siteId: String, request: UpdateSiteRequest): Site {
        val site = get(userId, siteId)

        request.name?.let { site.name = it }
        site.domain = request.domain
        site.url = request.url
        site.updatedAt = Instant.now()

        return siteRepository.save(site)
    }

    fun delete(userId: String, siteId: String) {
        val site = get(userId, siteId)

        site.deletedAt = Instant.now()
        site.updatedAt = Instant.now()

        siteRepository.save(site)
        siteKeyService.revokeActive(siteId)
    }

    fun keyInfo(userId: String, siteId: String): SiteKey {
        get(userId, siteId)

        return siteKeyRepository.findBySiteIdAndStatus(siteId, SiteKeyStatus.ACTIVE)
            ?: throw GlobalException.DataNotFound("Active site key not found")
    }

    fun rotateKey(userId: String, siteId: String): IssuedSiteKey {
        val site = get(userId, siteId)
        return siteKeyService.issue(site)
    }

    fun revokeKey(userId: String, siteId: String) {
        get(userId, siteId)
        siteKeyService.revokeActive(siteId)
    }
}